pkg/driver/subdir_provisioner.go

```diff
@@ -0,0 +1,173 @@
+			volumeContextMountName: mountname,
+			volumeContextBaseDir:   baseDir,
+		},
+	}); err != nil {
+		return nil, err
+	}
+
+	// Ensure to unmount dnsname@tcp:/mountname/baseDir at least once
+	defer func() {
+		if _, err := p.driver.NodeUnpublishVolume(ctx, &csi.NodeUnpublishVolumeRequest{
+			VolumeId:   tempUUID,
+			TargetPath: target,
+		}); err == nil {
+			if err := os.RemoveAll(target); err != nil {
+				klog.Warningf("Could not delete %q: %v", target, err)
+			}
+		} else {
+			klog.Warningf("Could not unmount %q: %v", target, err)
+		}
+	}()
+
+	// Create subDir (dnsname@tcp:/mountname/baseDir/subDir) under the tempMountPathPrefix/tempUUID
+	volName := req.GetName()
+	if err := p.driver.mounter.MakeDir(filepath.Join(target, volName)); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	return &csi.Volume{
+		VolumeId:      fmt.Sprintf("%s:%s:%s:%s:%s", dnsname, mountname, baseDir, volName, volName),
+		CapacityBytes: req.GetCapacityRange().GetRequiredBytes(),
+		VolumeContext: map[string]string{
+			volumeContextDnsName:   dnsname,
+			volumeContextMountName: mountname,
+			volumeContextBaseDir:   baseDir,
+			volumeContextSubDir:    volName,
+		},
+	}, nil
+}
+
+func (p SubDirProvisioner) Delete(ctx context.Context, req *csi.DeleteVolumeRequest) error {
+	volumeID := req.GetVolumeId()
+	fsxVolume, err := getFsxVolumeFromVolumeID(volumeID)
+	if err != nil {
+		return err
+	}
+
+	if len(fsxVolume.dnsname) == 0 {
+		return status.Error(codes.InvalidArgument, "dnsname is not provided")
+	}
+
+	if len(fsxVolume.mountname) == 0 {
+		return status.Error(codes.InvalidArgument, "mountname is not provided")
+	}
+
+	tempUUID := uuid.New().String()
+	target := filepath.Join(tempMountPathPrefix, tempUUID)
+
+	volCap := &csi.VolumeCapability{
+		AccessMode: &csi.VolumeCapability_AccessMode{
+			Mode: csi.VolumeCapability_AccessMode_SINGLE_NODE_WRITER,
+		},
+	}
+
+	// Mount dnsname@tcp:/mountname/baseDir to delete subDir
+	if _, err := p.driver.NodePublishVolume(ctx, &csi.NodePublishVolumeRequest{
+		VolumeId:         tempUUID,
+		TargetPath:       target,
+		VolumeCapability: volCap,
+		VolumeContext: map[string]string{
+			volumeContextDnsName:   fsxVolume.dnsname,
+			volumeContextMountName: fsxVolume.mountname,
+			volumeContextBaseDir:   fsxVolume.baseDir,
+		},
+	}); err != nil {
+		return err
+	}
+
+	// Ensure to unmount dnsname@tcp:/mountname/baseDir at least once
+	defer func() {
+		if _, err := p.driver.NodeUnpublishVolume(ctx, &csi.NodeUnpublishVolumeRequest{
+			VolumeId:   tempUUID,
+			TargetPath: target,
+		}); err == nil {
+			if err := os.RemoveAll(target); err != nil {
+				klog.Warningf("Could not delete %q: %v", target, err)
+			}
+		} else {
+			klog.Warningf("Could not unmount %q: %v", target, err)
+		}
+	}()
+
+	// Delete subDir (dnsname@tcp:/mountname/baseDir/subDir) under the tempMountPathPrefix/tempUUID
+	if err := os.RemoveAll(filepath.Join(target, fsxVolume.subDir)); err != nil {
```

Comment:
subDir and baseDir may need validation for relative paths. what happens if I set subDir to nonsense like ../